mysql-dst/mysql-cluster/man/ndb_blob_tool.1
'\" t
.\"     Title: \fBndb_blob_tool\fR
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 10/06/2018
.\"    Manual: MySQL Database System
.\"    Source: MySQL 5.7
.\"  Language: English
.\"
.TH "\FBNDB_BLOB_TOOL\FR" "1" "10/06/2018" "MySQL 5\&.7" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ndb_blob_tool \- check and repair BLOB and TEXT columns of NDB Cluster tables
.SH "SYNOPSIS"
.HP \w'\fBndb_blob_tool\ \fR\fB\fIoptions\fR\fR\fB\ \fR\fB\fItable\fR\fR\fB\ [\fR\fB\fIcolumn\fR\fR\fB,\ \&.\&.\&.]\fR\ 'u
\fBndb_blob_tool \fR\fB\fIoptions\fR\fR\fB \fR\fB\fItable\fR\fR\fB [\fR\fB\fIcolumn\fR\fR\fB, \&.\&.\&.]\fR
.SH "DESCRIPTION"
.PP
This tool can be used to check for and remove orphaned BLOB column parts from
NDB
tables, as well as to generate a file listing any orphaned parts\&. It is sometimes useful in diagnosing and repairing corrupted or damaged
NDB
tables containing
BLOB
or
TEXT
columns\&.
.PP
The basic syntax for
\fBndb_blob_tool\fR
is shown here:
.sp
.if n \{\
.RS 4
.\}
.nf
ndb_blob_tool [\fIoptions\fR] \fItable\fR [\fIcolumn\fR, \&.\&.\&.]
.fi
.if n \{\
.RE
.\}
.PP
Unless you use the
\fB\-\-help\fR
option, you must specify an action to be performed by including one or more of the options
\fB\-\-check\-orphans\fR,
\fB\-\-delete\-orphans\fR, or
\fB\-\-dump\-file\fR\&. These options cause
\fBndb_blob_tool\fR
to check for orphaned BLOB parts, remove any orphaned BLOB parts, and generate a dump file listing orphaned BLOB parts, respectively, and are described in more detail later in this section\&.
.PP
You must also specify the name of a table when invoking
\fBndb_blob_tool\fR\&. In addition, you can optionally follow the table name with the (comma\-separated) names of one or more
BLOB
or
TEXT
columns from that table\&. If no columns are listed, the tool works on all of the table\*(Aqs
BLOB
and
TEXT
columns\&. If you need to specify a database, use the
\fB\-\-database\fR
(\fB\-d\fR) option\&.
.PP
The
\fB\-\-verbose\fR
option provides additional information in the output about the tool\*(Aqs progress\&.
.PP
The following table includes options that are specific to
\fBndb_blob_tool\fR\&. Additional descriptions follow the table\&. For options common to most NDB Cluster programs (including
\fBndb_blob_tool\fR), see
Options Common to NDB Cluster Programs(1)\&.
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.B Table\ \&21.333.\ \&Command\-line options for the ndb_blob_tool program
.TS
allbox tab(:);
lB lB lB.
T{
Format
T}:T{
Description
T}:T{
Added, Deprecated, or Removed
T}
.T&
l l l
l l l
l l l
l l l
l l l.
T{
.PP
--check-orphans
T}:T{
Check for orphan blob parts
T}:T{
.PP
All MySQL 5.7 based releases
T}
T{
.PP
--database=db_name,
.PP
-d
T}:T{
Database to find the table in.
T}:T{
.PP
All MySQL 5.7 based releases
T}
T{
.PP
--delete-orphans
T}:T{
Delete orphan blob parts
T}:T{
.PP
All MySQL 5.7 based releases
T}
T{
.PP
--dump-file=file
T}:T{
Write orphan keys to specified file
T}:T{
.PP
All MySQL 5.7 based releases
T}
T{
.PP
--verbose,
.PP
-v
T}:T{
Verbose output
T}:T{
.PP
All MySQL 5.7 based releases
T}
.TE
.sp 1
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-check\-orphans\fR
.TS
allbox tab(:);
lB lB.
T{
Property
T}:T{
Value
T}
.T&
l l
l l
l l.
T{
\fBCommand-Line Format\fR
T}:T{
--check-orphans
T}
T{
\fBType\fR
T}:T{
Boolean
T}
T{
\fBDefault Value\fR
T}:T{
FALSE
T}
.TE
.sp 1
Check for orphaned BLOB parts in NDB Cluster tables\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-database=\fR\fB\fIdb_name\fR\fR,
\fB\-d\fR
.TS
allbox tab(:);
lB lB.
T{
Property
T}:T{
Value
T}
.T&
l l
l l
l l.
T{
\fBCommand-Line Format\fR
T}:T{
--database=db_name
T}
T{
\fBType\fR
T}:T{
String
T}
T{
\fBDefault Value\fR
T}:T{
[none]
T}
.TE
.sp 1
Specify the database to find the table in\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-delete\-orphans\fR
.TS
allbox tab(:);
lB lB.
T{
Property
T}:T{
Value
T}
.T&
l l
l l
l l.
T{
\fBCommand-Line Format\fR
T}:T{
--delete-orphans
T}
T{
\fBType\fR
T}:T{
Boolean
T}
T{
\fBDefault Value\fR
T}:T{
FALSE
T}
.TE
.sp 1
Remove orphaned BLOB parts from NDB Cluster tables\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-dump\-file=\fR\fB\fIfile\fR\fR
.TS
allbox tab(:);
lB lB.
T{
Property
T}:T{
Value
T}
.T&
l l
l l
l l.
T{
\fBCommand-Line Format\fR
T}:T{
--dump-file=file
T}
T{
\fBType\fR
T}:T{
File name
T}
T{
\fBDefault Value\fR
T}:T{
[none]
T}
.TE
.sp 1
Writes a list of orphaned BLOB column parts to
\fIfile\fR\&. The information written to the file includes the table key and BLOB part number for each orphaned BLOB part\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-verbose\fR
.TS
allbox tab(:);
lB lB.
T{
Property
T}:T{
Value
T}
.T&
l l
l l
l l.
T{
\fBCommand-Line Format\fR
T}:T{
--verbose
T}
T{
\fBType\fR
T}:T{
Boolean
T}
T{
\fBDefault Value\fR
T}:T{
FALSE
T}
.TE
.sp 1
Provide extra information in the tool\*(Aqs output regarding its progress\&.
.RE
Example.PP
First we create an
NDB
table in the
test
database, using the
CREATE TABLE
statement shown here:
.sp
.if n \{\
.RS 4
.\}
.nf
USE test;
CREATE TABLE btest (
    c0 BIGINT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
    c1 TEXT,
    c2 BLOB
)   ENGINE=NDB;
.fi
.if n \{\
.RE
.\}
.PP
Then we insert a few rows into this table, using a series of statements similar to this one:
.sp
.if n \{\
.RS 4
.\}
.nf
INSERT INTO btest VALUES (NULL, \*(Aqx\*(Aq, REPEAT(\*(Aqx\*(Aq, 1000));
.fi
.if n \{\
.RE
.\}
.PP
When run with
\fB\-\-check\-orphans\fR
against this table,
\fBndb_blob_tool\fR
generates the following output:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBndb_blob_tool \-\-check\-orphans \-\-verbose \-d test btest\fR
connected
processing 2 blobs
processing blob #0 c1 NDB$BLOB_19_1
NDB$BLOB_19_1: nextResult: res=1
total parts: 0
orphan parts: 0
processing blob #1 c2 NDB$BLOB_19_2
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=0
NDB$BLOB_19_2: nextResult: res=1
total parts: 10
orphan parts: 0
disconnected
NDBT_ProgramExit: 0 \- OK
.fi
.if n \{\
.RE
.\}
.PP
The tool reports that there are no
NDB
BLOB column parts associated with column
c1, even though
c1
is a
TEXT
column\&. This is due to the fact that, in an
NDB
table, only the first 256 bytes of a
BLOB
or
TEXT
column value are stored inline, and only the excess, if any, is stored separately; thus, if there are no values using more than 256 bytes in a given column of one of these types, no
BLOB
column parts are created by
NDB
for this column\&. See
Section\ \&11.8, \(lqData Type Storage Requirements\(rq, for more information\&.
.SH "COPYRIGHT"
.br
.PP
Copyright \(co 1997, 2018, Oracle and/or its affiliates. All rights reserved.
.PP
This documentation is free software; you can redistribute it and/or modify it only under the terms of the GNU General Public License as published by the Free Software Foundation; version 2 of the License.
.PP
This documentation is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with the program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA or see http://www.gnu.org/licenses/.
.sp
.SH "SEE ALSO"
For more information, please refer to the MySQL Reference Manual,
which may already be installed locally and which is also available
online at http://dev.mysql.com/doc/.
.SH AUTHOR
Oracle Corporation (http://dev.mysql.com/).
 

mysql-dst/mysql-cluster/ndb/tools/ndb_blob_tool.cpp
/* Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; version 2 of the License.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA */

#include <ndb_global.h>
#include <ndb_opts.h>

#include <NdbOut.hpp>
#include <NdbApi.hpp>
#include <NDBT.hpp>

static const char* opt_dbname = 0;
static my_bool opt_check_orphans = false;
static my_bool opt_delete_orphans = false;
static const char* opt_dump_file = 0;
static my_bool opt_verbose = false;

static FILE* g_dump_file = 0;
static FileOutputStream* g_dump_out = 0;
static NdbOut g_dump;

static Ndb_cluster_connection* g_ncc = 0;
static Ndb* g_ndb = 0;
static NdbDictionary::Dictionary* g_dic = 0;

static const char* g_tabname = 0;
static const NdbDictionary::Table* g_tab = 0;

struct Pk { // table pk
  const char* colname;
  Pk() {
    colname = 0;
  }
  ~Pk() {
    delete [] colname;
  }
};
static Pk* g_pklist = 0;
static int g_pkcount = 0;

struct Blob { // blob column and table
  int blobno;
  int colno;
  const char* colname;
  const char* blobname;
  const NdbDictionary::Table* blobtab;
  Blob() {
    blobno = -1;
    colno = -1;
    colname = 0;
    blobname = 0;
    blobtab = 0;
  }
  ~Blob() {
    delete [] colname;
    delete [] blobname;
  }
};
static Blob* g_bloblist = 0;
static int g_blobcount = 0;

static NdbTransaction* g_scantx = 0;
static NdbScanOperation* g_scanop = 0;

struct Val { // attr value scanned from blob
  const char* colname;
  NdbRecAttr* ra;
  Val() {
    colname = 0;
    ra = 0;
  }
  ~Val() {
    delete [] colname;
  }
};
static Val* g_vallist = 0;
static int g_valcount = 0;

#define CHK1(b) \
  if (!(b)) { \
    ret = -1; \
    break; \
  }

#define CHK2(b, e) \
  if (!(b)) { \
    g_err << "ERR: " << #b << " failed at line " << __LINE__ \
          << ": " << e << endl; \
    ret = -1; \
    break; \
  }

// re-inventing strdup
static const char*
newstr(const char* s)
{
  require(s != 0);
  char* s2 = new char [strlen(s) + 1];
  strcpy(s2, s);
  return s2;
}

static NdbError
getNdbError(Ndb_cluster_connection* ncc)
{
  NdbError err;
  err.code = g_ncc->get_latest_error();
  err.message = g_ncc->get_latest_error_msg();
  return err;
}

static int
doconnect()
{
  int ret = 0;
  do
  {
    g_ncc = new Ndb_cluster_connection(opt_ndb_connectstring);
    CHK2(g_ncc->connect(opt_connect_retries - 1, opt_connect_retry_delay) == 0, getNdbError(g_ncc));
    CHK2(g_ncc->wait_until_ready(30, 10) == 0, getNdbError(g_ncc));

    g_ndb = new Ndb(g_ncc, opt_dbname);
    CHK2(g_ndb->init() == 0, g_ndb->getNdbError());
    CHK2(g_ndb->waitUntilReady(30) == 0, g_ndb->getNdbError());
    g_dic = g_ndb->getDictionary();

    g_info << "connected" << endl;
  }
  while (0);
  return ret;
}

static void
dodisconnect()
{
  delete g_ndb;
  delete g_ncc;
  g_info << "disconnected" << endl;
}

static int
scanblobstart(const Blob& b)
{
  int ret = 0;

  do
  {
    require(g_scantx == 0);
    g_scantx = g_ndb->startTransaction();
    CHK2(g_scantx != 0, g_ndb->getNdbError());

    g_scanop = g_scantx->getNdbScanOperation(b.blobtab);
    CHK2(g_scanop != 0, g_scantx->getNdbError());

    const NdbOperation::LockMode lm = NdbOperation::LM_Exclusive;
    CHK2(g_scanop->readTuples(lm) == 0, g_scanop->getNdbError());

    for (int i = 0; i < g_valcount; i++)
    {
      Val& v = g_vallist[i];
      v.ra = g_scanop->getValue(v.colname);
      CHK2(v.ra != 0, v.colname << ": " << g_scanop->getNdbError());
    }
    CHK1(ret == 0);

    CHK2(g_scantx->execute(NoCommit) == 0, g_scantx->getNdbError());
  }
  while (0);

  return ret;
}

static int
scanblobnext(const Blob& b, int& res)
{
  int ret = 0;
  do
  {
    res = g_scanop->nextResult();
    CHK2(res == 0 || res == 1, g_scanop->getNdbError());
    g_info << b.blobname << ": nextResult: res=" << res << endl;
  }
  while (0);
  return ret;
}

static void
scanblobclose(const Blob& b)
{
  if (g_scantx != 0)
  {
    g_ndb->closeTransaction(g_scantx);
    g_scantx = 0;
  }
}

static int
checkorphan(const Blob&b, int& res)
{
  int ret = 0;

  NdbTransaction* tx = 0;
  NdbOperation* op = 0;

  do
  {
    tx = g_ndb->startTransaction();
    CHK2(tx != 0, g_ndb->getNdbError());

    op = tx->getNdbOperation(g_tab);
    CHK2(op != 0, tx->getNdbError());

    const NdbOperation::LockMode lm = NdbOperation::LM_Read;
    CHK2(op->readTuple(lm) == 0, op->getNdbError());

    for (int i = 0; i < g_pkcount; i++)
    {
      Val& v = g_vallist[i];
      require(v.ra != 0);
      require(v.ra->isNULL() == 0);
      const char* data = v.ra->aRef();
      CHK2(op->equal(v.colname, data) == 0, op->getNdbError());
    }
    CHK1(ret == 0);

    // read something to be safe
    NdbRecAttr* ra0 = op->getValue(g_vallist[0].colname);
    require(ra0 != 0);

    // not sure about the rules
    require(tx->getNdbError().code == 0);
    tx->execute(Commit);
    if (tx->getNdbError().code == 626)
    {
      g_info << "parent not found" << endl;
      res = 1; // not found
    }
    else
    {
      CHK2(tx->getNdbError().code == 0, tx->getNdbError());
      res = 0; // found
    }
  }
  while (0);

  if (tx != 0)
    g_ndb->closeTransaction(tx);
  return ret;
}

static int
deleteorphan(const Blob& b)
{
  int ret = 0;

  NdbTransaction* tx = 0;

  do
  {
    tx = g_ndb->startTransaction();
    CHK2(tx != 0, g_ndb->getNdbError());

    CHK2(g_scanop->deleteCurrentTuple(tx) == 0, g_scanop->getNdbError());
    CHK2(tx->execute(Commit) == 0, tx->getNdbError());
  }
  while (0);

  if (tx != 0)
    g_ndb->closeTransaction(tx);
  return ret;
}

static int
doorphan(const Blob& b)
{
  int ret = 0;
  do
  {
    g_err << "processing blob #" << b.blobno << " " << b.colname
          << " " << b.blobname << endl;

    if (opt_dump_file)
    {
      g_dump << "column: " << b.colname << endl;
      g_dump << "blob: " << b.blobname << endl;
      g_dump << "orphans (table key; blob part number):" << endl;
    }

    int totcount = 0;
    int orphancount = 0;

    CHK1(scanblobstart(b) == 0);
    while (1)
    {
      int res;
      res = -1;
      CHK1(scanblobnext(b, res) == 0);
      if (res != 0)
        break;
      totcount++;
      res = -1;
      CHK1(checkorphan(b, res) == 0);
      if (res != 0)
      {
        orphancount++;
        if (opt_dump_file)
        {
          g_dump << "key: ";
          for (int i = 0; i < g_valcount; i++)
          {
            const Val& v = g_vallist[i];
            g_dump << *v.ra;
            if (i + 1 < g_valcount)
              g_dump << ";";
          }
          g_dump << endl;
        }
        if (opt_delete_orphans)
        {
          CHK1(deleteorphan(b) == 0);
        }
      }
    }
    CHK1(ret == 0);

    g_err << "total parts: " << totcount << endl;
    g_err << "orphan parts: " << orphancount << endl;
    if (opt_dump_file)
    {
      g_dump << "total parts: " << totcount << endl;
      g_dump << "orphan parts: " << orphancount << endl;
    }
  }
  while (0);

  scanblobclose(b);
  return ret;
}

static int
doorphans()
{
  int ret = 0;
  g_err << "processing " << g_blobcount << " blobs" << endl;
  for (int i = 0; i < g_blobcount; i++)
  {
    const Blob& b = g_bloblist[i];
    CHK1(doorphan(b) == 0);
  }
  return ret;
}

static int
isblob(const NdbDictionary::Column* c)
{
  if (c->getType() == NdbDictionary::Column::Blob ||
      c->getType() == NdbDictionary::Column::Text)
    if (c->getPartSize() != 0)
      return 1;
  return 0;
}

static int
getobjs()
{
  int ret = 0;
  do
  {
    g_tab = g_dic->getTable(g_tabname);
    CHK2(g_tab != 0, g_tabname << ": " << g_dic->getNdbError());
    const int tabid = g_tab->getObjectId();
    const int ncol = g_tab->getNoOfColumns();

    g_pklist = new Pk [NDB_MAX_NO_OF_ATTRIBUTES_IN_KEY];
    for (int i = 0; i < ncol; i++)
    {
      const NdbDictionary::Column* c = g_tab->getColumn(i);
      if (c->getPrimaryKey())
      {
        Pk& p = g_pklist[g_pkcount++];
        const char* colname = c->getName();
        p.colname = newstr(colname);
      }
    }
    require(g_pkcount != 0 && g_pkcount == g_tab->getNoOfPrimaryKeys());

    g_valcount = g_pkcount + 1;
    g_vallist = new Val [g_valcount];
    for (int i = 0; i < g_valcount; i++)
    {
      Val& v = g_vallist[i];
      if (i < g_pkcount)
      {
        const Pk& p = g_pklist[i];
        v.colname = newstr(p.colname);
      }
      if (i == g_pkcount + 0) // first blob attr to scan
      {
        v.colname = newstr("NDB$PART");
      }
    }

    if (g_blobcount == 0)
    {
      for (int i = 0; i < ncol; i++)
      {
        const NdbDictionary::Column* c = g_tab->getColumn(i);
        if (isblob(c))
        {
          Blob& b = g_bloblist[g_blobcount++];
          const char* colname = c->getName();
          b.colname = newstr(colname);
        }
      }
    }

    for (int i = 0; i < g_blobcount; i++)
    {
      Blob& b = g_bloblist[i];
      b.blobno = i;
      const NdbDictionary::Column* c = g_tab->getColumn(b.colname);
      CHK2(c != 0, g_tabname << ": " << b.colname << ": no such column");
      CHK2(isblob(c), g_tabname << ": " << b.colname << ": not a blob");
      b.colno = c->getColumnNo();
      {
        char blobname[100];
        sprintf(blobname, "NDB$BLOB_%d_%d", tabid, b.colno);
        b.blobname = newstr(blobname);
      }
      b.blobtab = g_dic->getTable(b.blobname);
      CHK2(b.blobtab != 0, g_tabname << ": " << b.colname << ": " << b.blobname << ": " << g_dic->getNdbError());
    }
    CHK1(ret == 0);
  }
  while (0);
  return ret;
}

static int
doall()
{
  int ret = 0;
  do
  {
    if (opt_dump_file)
    {
      g_dump_file = fopen(opt_dump_file, "w");
      CHK2(g_dump_file != 0, opt_dump_file << ": " << strerror(errno));
      g_dump_out = new FileOutputStream(g_dump_file);
      new (&g_dump) NdbOut(*g_dump_out);

      const char* action = 0;
      if (opt_check_orphans)
        action = "check";
      if (opt_delete_orphans)
        action = "delete";

      g_dump << "table: " << g_tabname << endl;
      g_dump << "action: " << action << endl;
    }
    CHK1(doconnect() == 0);
    CHK1(getobjs() == 0);
    if (g_blobcount == 0)
    {
      g_err << g_tabname << ": no blob columns" << endl;
      break;
    }
    CHK1(doorphans() == 0);
  }
  while (0);

  dodisconnect();
  if (g_dump_file != 0)
  {
    g_dump << "result: "<< (ret == 0 ? "ok" : "failed") << endl;
    flush(g_dump);
    if (fclose(g_dump_file) != 0)
    {
      g_err << opt_dump_file << ": write failed: " << strerror(errno) << endl;
    }
    g_dump_file = 0;
  }
  return ret;
}

static struct my_option
my_long_options[] =
{
  NDB_STD_OPTS("ndb_blob_tool"),
  { "database", 'd',
    "Name of database table is in",
    (uchar**) &opt_dbname, (uchar**) &opt_dbname, 0,
    GET_STR, REQUIRED_ARG, 0, 0, 0, 0, 0, 0 },
  { "check-orphans", NDB_OPT_NOSHORT,
    "Check for orphan blob parts",
    (uchar **)&opt_check_orphans, (uchar **)&opt_check_orphans, 0,
    GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0 },
  { "delete-orphans", NDB_OPT_NOSHORT,
    "Delete orphan blob parts",
    (uchar **)&opt_delete_orphans, (uchar **)&opt_delete_orphans, 0,
    GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0 },
  { "dump-file", NDB_OPT_NOSHORT,
    "Write orphan keys (table key and part number) into file",
    (uchar **)&opt_dump_file, (uchar **)&opt_dump_file, 0,
    GET_STR, REQUIRED_ARG, 0, 0, 0, 0, 0, 0 },
  { "verbose", 'v',
    "Verbose messages",
    (uchar **)&opt_verbose, (uchar **)&opt_verbose, 0,
    GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0 },
  { 0, 0,
    0,
    0, 0, 0,
    GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0 }
};

const char*
load_default_groups[]= { "mysql_cluster", 0 };

static void
short_usage_sub()
{
  ndb_short_usage_sub("table [blobcolumn..]");
  printf("Default is to process all blob columns\n");
  printf("(1) Check orphans with --check --dump=out1.txt\n");
  printf("(2) Delete orphans with --delete --dump=out2.txt\n");
}

static void
usage()
{
  printf("%s: check and repair blobs\n", my_progname);
  ndb_usage(short_usage_sub, load_default_groups, my_long_options);
}

static int
checkopts(int argc, char** argv)
{
  if (opt_dbname == 0)
    opt_dbname = "TEST_DB";

  if (argc < 1)
  {
    g_err << "Table name required" << endl;
    return 1;
  }
  g_tabname = newstr(argv[0]);

  g_bloblist = new Blob [NDB_MAX_ATTRIBUTES_IN_TABLE];
  g_blobcount = argc - 1;
  for (int i = 0; i < g_blobcount; i++)
  {
    Blob& b = g_bloblist[i];
    b.colname = newstr(argv[1 + i]);
  }

  if (opt_check_orphans ||
      opt_delete_orphans)
  {
    if (opt_check_orphans &&
        opt_delete_orphans)
    {
      g_err << "Specify only one action (--check-orphans etc)" << endl;
      return 1;
    }
  }
  else
  {
    g_err << "Action (--check-orphans etc) required" << endl;
    return 1;
  }
  return 0;
}

static void
freeall()
{
  delete [] g_tabname;
  delete [] g_pklist;
  delete [] g_bloblist;
  delete [] g_vallist;

  delete g_dump_out;
  if (g_dump_file != 0)
    (void)fclose(g_dump_file);
}

int
main(int argc, char** argv)
{
  NDB_INIT("ndb_blob_tool");
  int ret;
  ndb_opt_set_usage_funcs(short_usage_sub, usage);
  ret = handle_options(&argc, &argv, my_long_options, ndb_std_get_one_option);
  if (ret != 0 || checkopts(argc, argv) != 0)
    return NDBT_ProgramExit(NDBT_WRONGARGS);

  setOutputLevel(opt_verbose ? 2 : 0);

  ret = doall();
  freeall();
  if (ret == -1)
    return NDBT_ProgramExit(NDBT_FAILED);
  return NDBT_ProgramExit(NDBT_OK);
}
 